addon/visit.js

```diff
@@ -170,27 +195,36 @@ class PageVisitor {
             } else if (e.message === 'Message manager disconnected') {
                 error = "tab unexpectedly closed (message manager disconnected)";
             }
-            this.setCurrentStatus({
-                message: 'freezing failed: ' + error, isError: true, isFinal: true
-            });
-        } finally {
-            clearTimeout(timer);
-            await browser.tabs.remove(tab.id);
+            this.errorAndStop(`freezing failed: ${error}`, tab.id, windowId);
         }
 
+        await browser.tabs.remove(tab.id);
         // Done with this url, trigger loading of the next.
         this.doc.dispatchEvent(new CustomEvent(
-            'fathom:next',
-            {detail: windowId}
+          'fathom:next',
+          {detail: windowId}
         ));
     }
 
+    // Do final cleanup and close the window.
+    async done(event) {
+        // If there was an error, we skip processing the results
+        if (event.detail.success) {
+            await this.processAtEndOfRun();
+        }
+        const windowId = event.detail.windowId;
+        browser.tabs.onUpdated.removeListener(this.tabUpdateListener);
+        this.tabUpdateListener = undefined;
+        browser.windows.remove(windowId);
+        this.doc.getElementById('freeze').disabled = false;
+    }
```

Comment:
This behavior happens for a bunch of errors: If the tab doesn't exist anymore (`Error: Invalid tab ID`), if there are any `null` feature vectors, if there are errors encountered during ruleset execution and for the `Receiving end does not exist` error).

One error situation where it is not caught is if we hit a 404. In this case, FathomFox happily vectorizes the 404 page (which is not the sample page) and moves on without mentioning anything. We should handle this situation, though I'm okay if it's a follow-up issue.

Unfortunately, we can't tell a 404 error from `browser.tabs.create` directly. Per rpl (a WE API Firefox developer):
> You should be able to achieve that by using a webRequest listener (a non-blocking one should be enough, and just listening for the main_frame ones using the filtering options, see https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/ResourceType) and use the tabId property in the webRequest event details to detect the ones related to the tab you are opening

Reply:
Also, I think we could be more descriptive with the error in the case that it's from ruleset execution. This is a similar tack to our custom `nullFeatures` check.

I spent almost an entire day yesterday trying to figure out why I couldn't get my Smoot articles samples vectorized in this PR, as the error logged in the Vectorizer page didn't pass along information about where the error occurred. It turned out I had unhandled errors for two sample pages in one of my rules. Perhaps the error message can be more descriptive if we know the error was thrown in executing a rule?

In this case, that'd be after we exceed `maxTries` attempts, if `vector.nodes` is `[]` (I haven't tested whether this is the case when we expect more than one candidate node, but this was the case in my situation for 'article'), we could prepend the error message with something like: "Encountered an error most likely in the ${traineeId} ruleset: ${error}." Though perhaps there's a better way to detect and communicate if a rule in the ruleset throws an error before returning a score?

Reply:
It seems reasonable to me to have Fathom itself catch errors thrown by scoring (and other) callbacks and re-raise them wrapped in more informative exceptions.

Reply:
> In this case, FathomFox happily vectorizes the 404 page

Just curious: are you really hitting 404s during vectorization, or do you mean corpus collection (since it's the same Visit framework)?

Reply:
Another thing to raise here:
In `master` for sure, and I'll have to check if this is the case here, a scoring callback in my ruleset errored and never returned. Those two samples for which that happened didn't show up at all in the resulting `vectors.json` file. We should ensure that we don't create a `vectors.json` file if we don't have a record of results for every sample in the list provided.

Reply:
> Just curious: are you really hitting 404s during vectorization, or do you mean corpus collection (since it's the same Visit framework)?

I was only testing Vectorization. I understand this kind of situation is uncommon -- particularly if the developer uses `fathom-list` to generate the list of paths to all the sample pages, however, in the process of debugging, a developer could still manually enter a path and have a typo, in which case, a 404 error would occur.

Reply:
I'd like to make sure I follow all of this. There seem to be three issues:

1. We don't have a way of detecting if a page 404's. If a 404 error occurs, the vectorizer runs on the error page.
2. If a rule throws an error, the `vector` object returned by the `vectorizeTab` message will be `[]`, which satisfies the `while` loop's check against undefined and does not trigger a null vector error. We could handle the error within Fathom an re-raising an informative exception that `processWithinTimeout()` could handle.
3. If the situation in (2) happens, there will be no element in `vectors.json` for that page. We should make sure we have a element in `this.vectors` for every page before writing a `vectors.json` file.

Reply:
1. Correct.
2. That was true in my case for 'article', though in that situation we only have likely one candidate element. I'm not sure if this is true generally, say if we had many candidate elements on the page, and the rule threw an error after we had already processed some others without issue. That needs checking. Whether we check against `vectors.nodes` being empty or not, we should be able to tell when an error is thrown in a rule and pass additional information that the error is coming from the ruleset.
3. This was true when I used FathomFox from AMO, although I don't see the same behavior in this PR, as I just tried it again. On this branch, a `vectors.json` file is _not_ created when my two samples throw an error inside a Fathom rule, so I think just making sure that error gets passed along and is more descriptive re: where it failed is sufficient. Please do confirm this though! If you can think of other situations where this might happen, then it makes sense to have that check. I'd have to think about it more.